src/lib/math/mp/mp_asmi.h

```diff
@@ -11,115 +11,138 @@
 
 #include <botan/types.h>
 
-#if BOTAN_MP_WORD_BITS == 64
+#if !defined(BOTAN_TARGET_HAS_NATIVE_UINT128)
    #include <botan/internal/donna128.h>
-   #include <botan/internal/mul128.h>
 #endif
 
 namespace Botan {
 
-#if BOTAN_MP_WORD_BITS == 32
-typedef uint64_t dword;
-   #define BOTAN_HAS_NATIVE_DWORD
-
-#elif BOTAN_MP_WORD_BITS == 64
-   #if defined(BOTAN_TARGET_HAS_NATIVE_UINT128)
-typedef uint128_t dword;
-      #define BOTAN_HAS_NATIVE_DWORD
-   #else
-typedef donna128 dword;
-   #endif
-
-#else
-   #error BOTAN_MP_WORD_BITS must be 32 or 64
-#endif
-
 #if defined(BOTAN_USE_GCC_INLINE_ASM)
-
-   #if defined(BOTAN_TARGET_ARCH_IS_X86_32) && (BOTAN_MP_WORD_BITS == 32)
+   #if defined(BOTAN_TARGET_ARCH_IS_X86_FAMILY)
       #define BOTAN_MP_USE_X86_32_ASM
-   #elif defined(BOTAN_TARGET_ARCH_IS_X86_64) && (BOTAN_MP_WORD_BITS == 64)
+   #endif
+
+   #if defined(BOTAN_TARGET_ARCH_IS_X86_64)
       #define BOTAN_MP_USE_X86_64_ASM
    #endif
+#endif
 
+/*
+* Concept for allowed multiprecision word types
+*/
+template <typename T>
+concept WordType = (std::same_as<T, uint32_t> || std::same_as<T, uint64_t>);
+
+template <WordType W>
+struct DwordType {};
+
+template <>
+struct DwordType<uint32_t> {
+   public:
+      typedef uint64_t type;
+      static const bool is_native = true;
+};
```

Comment:
Nice! I can actually use this in my CRYSTALS refactoring, following up on #3887.